lib/segment/src/index/hnsw_index/graph_layers.rs

```diff
@@ -149,21 +149,27 @@ impl GraphLayers {
         visited_list: &mut VisitedList,
         points_scorer: &FilteredScorer,
     ) {
+        let limit = self.get_m(level);
+        let mut scores_buffer: Vec<ScoredPointOffset> =
+            vec![ScoredPointOffset { idx: 0, score: 0. }; limit];
+        let mut points_ids: Vec<PointOffsetType> = vec![];
+        points_ids.reserve(2 * limit);
+
         while let Some(candidate) = searcher.candidates.pop() {
             if candidate.score < searcher.lower_bound() {
                 break;
             }
-            let mut links_iter = self
-                .links(candidate.idx, level)
-                .iter()
-                .copied()
-                .filter(|point_id| !visited_list.check_and_update_visited(*point_id));
-
-            points_scorer.score_iterable_points(
-                &mut links_iter,
-                self.get_m(level),
-                |score_point| searcher.process_candidate(score_point),
-            );
+
+            points_ids.clear();
+            for link in self.links(candidate.idx, level) {
+                if !visited_list.check_and_update_visited(*link) {
```

Comment:
In theory the amount of edges for a single node could be larger than `get_m` - in case we are creating additional links based in indexed payload. In this case we need to limit scoring to only the first `M` points which pass the filter, even if there are more than `M` point in total